pkg/workflow/workflowrun/operator.go

```diff
@@ -454,10 +456,37 @@ func (o *operator) GC(lastTry, wfrDeletion bool) error {
 				o.recorder.Eventf(o.wfr, corev1.EventTypeWarning, "GC", "Delete pod '%s' error: %v", status.Pod.Name, err)
 			}
 		} else {
-			log.WithField("ns", status.Pod.Namespace).WithField("pod", status.Pod.Name).Info("Pod deleted")
+			log.WithField("ns", status.Pod.Namespace).WithField("pod", status.Pod.Name).Info("Start to delete pod")
+
+			wg.Add(1)
+			go func(namespace, podName string) {
+				defer wg.Done()
+
+				timeout := time.After(5 * time.Minute)
+				ticker := time.NewTicker(5 * time.Second)
+				defer ticker.Stop()
+				for {
+					select {
+					case <-timeout:
+						log.WithField("ns", namespace).WithField("pod", podName).Warn("Pod deletion timeout")
+						return
+					case <-ticker.C:
+						_, err := o.clusterClient.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+						if err != nil && errors.IsNotFound(err) {
+							log.WithField("ns", namespace).WithField("pod", podName).Info("Pod deleted")
+							return
+						}
+					}
+				}
+			}(status.Pod.Namespace, status.Pod.Name)
 		}
 	}
 
+	// Wait all workflowRun related workload pods deleting completed, then start gc pod to clean date on PV.
+	// Otherwise, if the path which is used by workload pods in the PV is deleted before workload pods deletion,
+	// the pod deletion process will get stuck on Terminating status.
+	wg.Wait()
```

Comment:
does this mean that if a workflowrun pod stuck on terminating for some reason, gc would pause for 5 min?

Reply:
Yes，I don't know how to judge the error more precisely.

Reply:
for every finished wfr pod, a new gc pod would run to clean its subpath?

Reply:
One wfr may start multiple pods(one stage a pod), after all pods started by the wfr finished, a new gc pod would run to clean data.

Reply:
so if any of the pods started by a wfr stuck on terminating for more than 5 min, the gc for this wfr will fail and the subpath will remain in the pv indefinitely(since we don't have any way of triggering gc for this particular wfr again)?

Reply:
No, 5 min is a timeout threshold, if timeout, cyclone will start a  gc pod to delete subpath.

Reply:
got it. so this does not solve the issue once and for all. it fixes the issue for most of the scenarios while we may still expect glitches in some corner cases?

Reply:
Right, Do you think it's acceptable?

Reply:
I think this might be the best we can get with minimal effort. And in cases where wfr pods stuck on terminating for other reasons, chances are we have to handle it manually(and it is stuck anyways).

Just off the top of my head, is it possible if we have the gc run periodically and check all of the subpath in the pvc and delete those which does not have a corresponding pod(suppose we have a one to one mapping between pod name and subpath name) so that we will never delete any subpath before pod cleanup? Not sure if this way we can solve the issue. Anyways, this is not a trivial change that is appropriate for a patch version. 

So I think this is good enough for now, but hopefully we can come up with a better solution that will eliminate the issue once and for all.

Reply:
SGTM, and I submit [another commit](https://github.com/caicloud/cyclone/pull/1321/commits/fb2cea27f8cc2ac06e75dac0c13957c6384903e1) to ensure that **only the last** retry GC process (cyclone GC, deleting stage pods and deleting data in PV, has a retry mechanism) will delete data in PV even if there are stage pods not finished, otherwise, cyclone will wait for the next retry GC to execute. 

Although this commit can not eliminate the issue once and for all, It will reduce the probability of the corner case occurrences.

Reply:
@pendoragon PTAL

Reply:
LGTM